File: express/src/routes/Content/article_comments.ts
import { Router } from 'express';
import { prisma } from '../../lib/prisma.js';
import { ArticleCommentService } from '../../services/Content/article_comment_service.js';

const router = Router();
const service = new ArticleCommentService();


router.get('/', async (_req, res) => {
  const items = await prisma.articleComment.findMany();
  res.json(items);
});

router.post('/', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.body !== undefined) data.body = body.body;
    if (body.isHidden !== undefined) data.isHidden = body.isHidden;
    if (body.createdAt !== undefined) data.createdAt = body.createdAt != null ? new Date(body.createdAt) : null;
    if (body.articleId !== undefined) data.articleId = body.articleId;
    if (body.authorId !== undefined) data.authorId = body.authorId;
    if (body.parentCommentId !== undefined) data.parentCommentId = body.parentCommentId;
  try {
    const entity = await prisma.articleComment.create({ data });
    res.status(201).json(entity);
  } catch (err: any) {
    res.status(400).json({ error: err?.message ?? 'Validation error' });
  }
});

router.get('/:id', async (req, res) => {
  const entity = await prisma.articleComment.findUnique({ where: { id: Number(req.params.id) } });
  if (!entity) return res.status(404).json({ error: 'Not found' });
  res.json(entity);
});

router.put('/:id', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.body !== undefined) data.body = body.body;
    if (body.isHidden !== undefined) data.isHidden = body.isHidden;
    if (body.createdAt !== undefined) data.createdAt = body.createdAt != null ? new Date(body.createdAt) : null;
    if (body.articleId !== undefined) data.articleId = body.articleId;
    if (body.authorId !== undefined) data.authorId = body.authorId;
    if (body.parentCommentId !== undefined) data.parentCommentId = body.parentCommentId;
  try {
    const entity = await prisma.articleComment.update({ where: { id: Number(req.params.id) }, data });
    res.json(entity);
  } catch (err: any) {
    const status = err?.code === 'P2025' ? 404 : 400;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.patch('/:id', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.body !== undefined) data.body = body.body;
    if (body.isHidden !== undefined) data.isHidden = body.isHidden;
    if (body.createdAt !== undefined) data.createdAt = body.createdAt != null ? new Date(body.createdAt) : null;
    if (body.articleId !== undefined) data.articleId = body.articleId;
    if (body.authorId !== undefined) data.authorId = body.authorId;
    if (body.parentCommentId !== undefined) data.parentCommentId = body.parentCommentId;
  try {
    const entity = await prisma.articleComment.update({ where: { id: Number(req.params.id) }, data });
    res.json(entity);
  } catch (err: any) {
    const status = err?.code === 'P2025' ? 404 : 400;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.delete('/:id', async (req, res) => {
  try {
    await prisma.articleComment.delete({ where: { id: Number(req.params.id) } });
    res.status(204).send();
  } catch {
    res.status(404).json({ error: 'Not found' });
  }
});

router.post('/:id/hide', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.hide(id);
    res.status(204).send();
  } catch (err: any) {
    res.status(404).json({ error: err?.message ?? 'Not found' });
  }
});

router.post('/:id/unhide', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.unhide(id);
    res.status(204).send();
  } catch (err: any) {
    res.status(404).json({ error: err?.message ?? 'Not found' });
  }
});
export default router;

File: gin/internal/handler/content/article_comments.go
package handler_content

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	model "cards_project/internal/model/content"
	"cards_project/internal/handler"
)

type ArticleCommentHandler struct { db *gorm.DB }

func NewArticleCommentHandler(db *gorm.DB) *ArticleCommentHandler {
	return &ArticleCommentHandler{db: db}
}

func (h *ArticleCommentHandler) RegisterRoutes(r gin.IRouter) {
	g := r.Group("/api/article_comments")
	g.GET("", h.List)
	g.POST("", h.Create)
	g.GET("/:id", h.Get)
	g.PUT("/:id", h.Update)
	g.PATCH("/:id", h.Patch)
	g.DELETE("/:id", h.Delete)
	g.POST("/:id/api/comments/{id}/hide", h.Hide)
	g.POST("/:id/api/comments/{id}/unhide", h.Unhide)
}

func (h *ArticleCommentHandler) List(c *gin.Context) {
	skip, limit := handler.Paginate(c)
	var rows []model.ArticleComment
	if err := h.db.Offset(skip).Limit(limit).Find(&rows).Error; err != nil {
		handler.DbError(c, err); return
	}
	out := make([]model.ArticleCommentResponse, len(rows))
	for i := range rows { out[i] = rows[i].ToResponse() }
	c.JSON(http.StatusOK, out)
}

func (h *ArticleCommentHandler) Create(c *gin.Context) {
	var req model.ArticleCommentCreateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	row := model.ArticleComment{}
	row.Body = req.Body
	row.IsHidden = req.IsHidden
	row.ArticleID = req.ArticleID
	row.AuthorID = req.AuthorID
	row.ParentCommentID = req.ParentCommentID
	if err := h.db.Create(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusCreated, row.ToResponse())
}

func (h *ArticleCommentHandler) Get(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.ArticleComment
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "ArticleComment"); return }
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *ArticleCommentHandler) Update(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.ArticleComment
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "ArticleComment"); return }
		handler.DbError(c, err); return
	}
	var req model.ArticleCommentUpdateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	row.ApplyUpdate(req)
	if err := h.db.Save(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *ArticleCommentHandler) Patch(c *gin.Context) { h.Update(c) }

func (h *ArticleCommentHandler) Delete(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	if err := h.db.Delete(&model.ArticleComment{}, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "ArticleComment"); return }
		handler.DbError(c, err); return
	}
	c.Status(http.StatusNoContent)
}

func (h *ArticleCommentHandler) Hide(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.ArticleComment
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "ArticleComment"); return }
		handler.DbError(c, err); return
	}
	err := row.Hide()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *ArticleCommentHandler) Unhide(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.ArticleComment
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "ArticleComment"); return }
		handler.DbError(c, err); return
	}
	err := row.Unhide()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}
